### PR TITLE
[Menu] Fix for keyboard focus is lost when hovering out of a menu

### PR DESCRIPTION
**Pull request by chiuam**
### Platforms Impacted
- [ ] iOS
- [X] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

~~The proposed fix is to let FocusZone handle the removal of focus by calling `resignFirstResponder` on the focused element. That requires creating a tracking area for the FocusZone window and overriding `mouseExited`.~~

~~Followup: In the after video, you can see that pressing the down arrow key sets keyboard focus on the subsequent item of the last focused item, where it should be the top item (last item w/ up arrow key).~~

Edit: A better fix was suggested by @Saadnajmi, that is to set focus on `onMouseLeave`, a simple one line fix :)

### Verification

Before:
  

https://user-images.githubusercontent.com/67026167/227336554-285357c8-ebf2-41ce-ba85-747eaea79e15.mov

After:

https://user-images.githubusercontent.com/67026167/227336582-de66cc6a-478c-4790-85ca-ceea42e333ca.mov

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [X] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 1]
```diff
@@ -186,6 +189,39 @@ static BOOL ShouldSkipFocusZone(NSView *view)
...
+
+- (void)mouseExited:(NSEvent *)event
+{
+	[super mouseExited:event];
+	NSView *viewWithFocus = GetFirstResponder([self window]);
+	if (viewWithFocus != nil)
+	{
+		[viewWithFocus resignFirstResponder];
+	}
+}
```
Comment (Saadnajmi) [inline packages/components/FocusZone/macos/RCTFocusZone.m:223]: This would affect all consumers of FocusZone, not just menu right? Would this negatively impact stuff like Radiogroup?

Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m:223]: That's right, I did a quick test on RadioGroup and Tabs before publishing the PR, didn't find anything weird... will do some more testing and report back

Comment (harrieshin) [reply packages/components/FocusZone/macos/RCTFocusZone.m:223]: just to be sure, lets make sure we test various edge cases we have been fixing over the years in focus zone so we aren't regressing anything.

File: packages/components/Menu/src/MenuItem/useMenuItem.ts [hunk 2]
```diff
@@ -104,7 +104,11 @@ export const useHoverFocusEffect = (hovered: boolean, componentRef: React.Mutabl
     if (hovered) {
       componentRef?.current?.focus();
     } else {
-      componentRef?.current?.blur();
+      // Since MenuList uses FocusZone on macOS, we need to handle the .blur() logic there in order to
```
Comment (harrieshin) [inline packages/components/Menu/src/MenuItem/useMenuItem.ts:107]: for my own understanding? why does blur logic interfere with keyboard navigation?

Comment (Saadnajmi) [reply packages/components/Menu/src/MenuItem/useMenuItem.ts:107]: When we blur on mouseExit, focus is lost from the menu overall. Then, the user can't ArrowUp/Down back between menu items until they put their mouse back over the menu.

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 1 above)
Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m:223]: Debugged Tabs and RadioGroup in Xcode, I saw that `[viewWithFocus resignFirstResponder];` does get called on mouseExited. I think this is fine because it only removes the highlight, not the focus ring that those controls use.

source: https://developer.apple.com/documentation/uikit/uiresponder/1621097-resignfirstresponder

Comment (harrieshin) [reply packages/components/FocusZone/macos/RCTFocusZone.m:223]: that link is for uikit though. not app kit

File: packages/components/Menu/src/MenuItem/useMenuItem.ts (hunk 2 above)
Comment (harrieshin) [reply packages/components/Menu/src/MenuItem/useMenuItem.ts:107]: can we make sure we capture that in comment and the pr description please?

**Review by harrieshin — APPROVED** [review]
sign off with comments

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 1 above)
Comment (harrieshin) [reply packages/components/FocusZone/macos/RCTFocusZone.m:223]: that is a good point. in the radio group, if it removes the highlight? what is focused? then?

Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m]: Scratch all that, a much simpler fix is to set focus on the menu container when onMouseLeave is called. Sorry for the noise :/

File: packages/components/Menu/src/MenuList/MenuList.tsx [hunk 3]
```diff
@@ -83,7 +83,7 @@ export const MenuList = compose<MenuListType>({
 
       const content =
         Platform.OS === 'macos' ? (
-          <Slots.root>
+          <Slots.root onMouseLeave={() => focusZoneRef.current.focus()}>
```
Comment (Saadnajmi) [inline packages/components/Menu/src/MenuList/MenuList.tsx:86]: Might be good to extract this to a a separate function / React.useCallback because we try to not pass lamba functions directly in jsx.

**Review by Saadnajmi — APPROVED** [review]

File: change/@fluentui-react-native-focus-zone-bb03f1cd-8bb5-4100-a714-2ae966d3310c.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Fix for keyboard focus is lost when hovering out of a menu",
+  "packageName": "@fluentui-react-native/focus-zone",
```
Comment (Saadnajmi) [inline change/@fluentui-react-native-focus-zone-bb03f1cd-8bb5-4100-a714-2ae966d3310c.json:4]: We can delete this file?

**Review by Saadnajmi — APPROVED** [review]

**Review by Saadnajmi — APPROVED** [review]
